src/tike/ptycho/solvers/combined.py

```diff
@@ -33,42 +34,50 @@ def cgrad(
 
     """
     cost = np.inf
-
-    if recover_psi:
-        psi, cost = _update_object(
-            op,
-            comm,
-            data,
-            psi,
-            scan,
-            probe,
-            num_iter=cg_iter,
-            step_length=step_length,
-        )
-
-    if recover_probe:
-        probe, cost = _update_probe(
-            op,
-            comm,
-            data,
-            psi,
-            scan,
-            probe,
-            num_iter=cg_iter,
-            step_length=step_length,
-            probe_is_orthogonal=probe_is_orthogonal,
-            mode=list(range(probe[0].shape[-3])),
-        )
-
-    if recover_positions and comm.pool.num_workers == 1:
-        scan, cost = update_positions_pd(
-            op,
-            comm.pool.gather(data, axis=1),
-            psi[0],
-            probe[0],
-            comm.pool.gather(scan, axis=1),
-        )
-        scan = comm.pool.bcast(scan)
+    # Unique batch for each device
+    batches = [
+        batch_indicies(s.shape[-2], num_batch, subset_is_random) for s in scan
+    ]
+    for n in range(num_batch):
+
+        bdata = comm.pool.map(collect_batch, data, batches, n=n)
```

Comment:
Is a `batch` equal to an item in `data` list generated by `split_by_scan_grid`? If so, is the purpose of this statement to load the batch to different GPUs?

Reply:
First the data is split position into a grid; this is when it is pushed to the devices because each device gets one part of the grid. Then each of these batches is randomly selected from the data on each device. No data is loaded from the host to device in the solvers module.

Reply:
Could you remind me of the benefit of using the batch-based approach? If my understanding is correct, each device has its own portion of data, then you split the piece of data into batches and process them in random order. As all local data in a device can be handle simultaneously, won't the batching approach decrease the overall performance?

Reply:
There are two benefits:
1. You **don't** always have enough working space to process all of the data simultaneously.
2. If we assume that a small batch of data can approximate the true gradient, then we can use small batches reduce the amount of work performed per update (step).

Reply:
OK, thanks for clarifying it. You can go ahead and merge it.